File: marksclock/routers/converters.py
"""Converter APIs — timezone, unix, ISO 8601, duration, date math."""

from __future__ import annotations

from datetime import datetime, timezone, timedelta
from zoneinfo import ZoneInfo

from dateutil.relativedelta import relativedelta
from fastapi import APIRouter
from pydantic import BaseModel

router = APIRouter()


# --- Timezone conversion ---

class TzConvert(BaseModel):
    time_iso: str
    from_tz: str
    to_tz: str


@router.post("/timezone")
async def convert_timezone(body: TzConvert) -> dict:
    dt = datetime.fromisoformat(body.time_iso)
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=ZoneInfo(body.from_tz))
    result = dt.astimezone(ZoneInfo(body.to_tz))
    return {
        "from": {"timezone": body.from_tz, "time": dt.isoformat()},
        "to": {"timezone": body.to_tz, "time": result.isoformat()},
    }


# --- Unix timestamp ---

class UnixConvert(BaseModel):
    timestamp: float | None = None
    iso: str | None = None


@router.post("/unix")
async def convert_unix(body: UnixConvert) -> dict:
    if body.timestamp is not None:
        dt = datetime.fromtimestamp(body.timestamp, tz=timezone.utc)
        return {"timestamp": body.timestamp, "iso": dt.isoformat(), "human": dt.strftime("%Y-%m-%d %H:%M:%S UTC")}
    elif body.iso:
        dt = datetime.fromisoformat(body.iso)
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return {"timestamp": dt.timestamp(), "iso": dt.isoformat(), "human": dt.strftime("%Y-%m-%d %H:%M:%S %Z")}
    return {"error": "provide timestamp or iso"}


# --- ISO 8601 ---

class IsoConvert(BaseModel):
    input: str


@router.post("/iso")
async def convert_iso(body: IsoConvert) -> dict:
    try:
        dt = datetime.fromisoformat(body.input)
        return {
            "iso": dt.isoformat(),
            "date": dt.strftime("%Y-%m-%d"),
            "time": dt.strftime("%H:%M:%S"),
            "timezone": str(dt.tzinfo) if dt.tzinfo else "naive",
            "epoch": dt.timestamp() if dt.tzinfo else None,
        }
    except ValueError:
        return {"error": f"Cannot parse: {body.input}"}


# --- Duration / time difference ---

class DurationCalc(BaseModel):
    start: str  # ISO
    end: str    # ISO


@router.post("/duration")
async def calc_duration(body: DurationCalc) -> dict:
    d1 = datetime.fromisoformat(body.start)
    d2 = datetime.fromisoformat(body.end)
    delta = d2 - d1
    total_seconds = delta.total_seconds()
    days = int(total_seconds // 86400)
    hours = int((total_seconds % 86400) // 3600)
    minutes = int((total_seconds % 3600) // 60)
    seconds = int(total_seconds % 60)
    return {
        "total_seconds": total_seconds,
        "days": days, "hours": hours, "minutes": minutes, "seconds": seconds,
        "human": f"{days}d {hours}h {minutes}m {seconds}s",
    }


# --- Date add/subtract ---

class DateAdd(BaseModel):
    date: str  # ISO date or datetime
    years: int = 0
    months: int = 0
    days: int = 0
    hours: int = 0
    minutes: int = 0
    seconds: int = 0


@router.post("/date-add")
async def date_add(body: DateAdd) -> dict:
    dt = datetime.fromisoformat(body.date)
    result = dt + relativedelta(
        years=body.years, months=body.months, days=body.days,
        hours=body.hours, minutes=body.minutes, seconds=body.seconds,
    )
    return {"original": dt.isoformat(), "result": result.isoformat()}


# --- Date utilities ---

@router.get("/day-of-week")
async def day_of_week(date: str) -> dict:
    dt = datetime.fromisoformat(date)
    return {"date": date, "day_of_week": dt.strftime("%A"), "day_number": dt.weekday()}


@router.get("/week-number")
async def week_number(date: str) -> dict:
    dt = datetime.fromisoformat(date)
    iso_cal = dt.isocalendar()
    return {"date": date, "iso_week": iso_cal[1], "iso_year": iso_cal[0], "iso_day": iso_cal[2]}


@router.get("/leap-year")
async def leap_year(year: int) -> dict:
    is_leap = (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0)
    return {"year": year, "is_leap_year": is_leap}


@router.get("/age")
async def age(birthdate: str) -> dict:
    birth = datetime.fromisoformat(birthdate).date()
    today = datetime.now().date()
    rd = relativedelta(today, birth)
    total_days = (today - birth).days
    return {
        "birthdate": str(birth),
        "years": rd.years, "months": rd.months, "days": rd.days,
        "total_days": total_days,
        "human": f"{rd.years} years, {rd.months} months, {rd.days} days",
    }


@router.get("/days-between")
async def days_between(start: str, end: str) -> dict:
    d1 = datetime.fromisoformat(start).date()
    d2 = datetime.fromisoformat(end).date()
    delta = (d2 - d1).days
    return {"start": str(d1), "end": str(d2), "days": delta, "weeks": delta / 7}

File: marksclock/static/js/converters.js
// Converters / calculators page

import { api, $ } from './utils.js';

export async function render(container) {
  container.innerHTML = `
    <h2 class="page-title">Time Tools</h2>

    <div class="card">
      <h3 class="card-title mb-1">Timezone Converter</h3>
      <div class="form-row">
        <div class="form-group"><label class="form-label">Date/Time</label><input class="input" id="tzTime" type="datetime-local" /></div>
        <div class="form-group"><label class="form-label">From</label><input class="input" id="tzFrom" value="America/New_York" list="tzList1" /></div>
        <div class="form-group"><label class="form-label">To</label><input class="input" id="tzTo" value="Europe/London" list="tzList1" /></div>
      </div>
      <button class="btn btn-primary" id="tzConvert">Convert</button>
      <div id="tzResult" class="mt-1 mono"></div>
      <datalist id="tzList1"></datalist>
    </div>

    <div class="card">
      <h3 class="card-title mb-1">Unix Timestamp</h3>
      <div class="form-row">
        <div class="form-group"><label class="form-label">Timestamp</label><input class="input" id="unixTs" type="number" placeholder="e.g. 1700000000" /></div>
        <div class="form-group"><label class="form-label">Or ISO string</label><input class="input" id="unixIso" placeholder="2024-01-01T00:00:00Z" /></div>
      </div>
      <button class="btn btn-primary" id="unixConvert">Convert</button>
      <div id="unixResult" class="mt-1 mono"></div>
    </div>

    <div class="card">
      <h3 class="card-title mb-1">Time Difference</h3>
      <div class="form-row">
        <div class="form-group"><label class="form-label">Start</label><input class="input" id="diffStart" type="datetime-local" /></div>
        <div class="form-group"><label class="form-label">End</label><input class="input" id="diffEnd" type="datetime-local" /></div>
      </div>
      <button class="btn btn-primary" id="diffCalc">Calculate</button>
      <div id="diffResult" class="mt-1 mono"></div>
    </div>

    <div class="card">
      <h3 class="card-title mb-1">Date Add/Subtract</h3>
      <div class="form-row">
        <div class="form-group"><label class="form-label">Start Date</label><input class="input" id="addDate" type="date" /></div>
        <div class="form-group"><label class="form-label">Years</label><input class="input" id="addY" type="number" value="0" /></div>
        <div class="form-group"><label class="form-label">Months</label><input class="input" id="addM" type="number" value="0" /></div>
        <div class="form-group"><label class="form-label">Days</label><input class="input" id="addD" type="number" value="0" /></div>
      </div>
      <button class="btn btn-primary" id="addCalc">Calculate</button>
      <div id="addResult" class="mt-1 mono"></div>
    </div>

    <div class="grid-2">
      <div class="card">
        <h3 class="card-title mb-1">Age Calculator</h3>
        <div class="form-group"><label class="form-label">Birthdate</label><input class="input" id="ageBirth" type="date" /></div>
        <button class="btn btn-primary" id="ageCalc">Calculate</button>
        <div id="ageResult" class="mt-1 mono"></div>
      </div>

      <div class="card">
        <h3 class="card-title mb-1">Days Between</h3>
        <div class="form-group"><label class="form-label">Start</label><input class="input" id="betStart" type="date" /></div>
        <div class="form-group"><label class="form-label">End</label><input class="input" id="betEnd" type="date" /></div>
        <button class="btn btn-primary" id="betCalc">Calculate</button>
        <div id="betResult" class="mt-1 mono"></div>
      </div>

      <div class="card">
        <h3 class="card-title mb-1">Leap Year</h3>
        <div class="form-group"><label class="form-label">Year</label><input class="input" id="leapYear" type="number" value="${new Date().getFullYear()}" /></div>
        <button class="btn btn-primary" id="leapCalc">Check</button>
        <div id="leapResult" class="mt-1 mono"></div>
      </div>

      <div class="card">
        <h3 class="card-title mb-1">Day of Week</h3>
        <div class="form-group"><label class="form-label">Date</label><input class="input" id="dowDate" type="date" /></div>
        <button class="btn btn-primary" id="dowCalc">Check</button>
        <div id="dowResult" class="mt-1 mono"></div>
      </div>
    </div>
  `;

  // Load timezone list
  const tzData = await api('/reference/timezones');
  const dl = $('#tzList1');
  for (const tz of tzData) {
    const opt = document.createElement('option');
    opt.value = tz.timezone;
    dl.appendChild(opt);
  }

  // Set defaults
  const now = new Date();
  const localIso = now.toISOString().slice(0, 16);
  $('#tzTime').value = localIso;
  $('#diffStart').value = localIso;
  $('#diffEnd').value = localIso;
  $('#addDate').value = now.toISOString().slice(0, 10);
  $('#ageBirth').value = '1990-01-01';
  $('#betStart').value = now.toISOString().slice(0, 10);
  $('#betEnd').value = now.toISOString().slice(0, 10);
  $('#dowDate').value = now.toISOString().slice(0, 10);

  // Timezone convert
  $('#tzConvert').onclick = async () => {
    const r = await api('/convert/timezone', { method: 'POST', body: { time_iso: $('#tzTime').value, from_tz: $('#tzFrom').value, to_tz: $('#tzTo').value } });
    $('#tzResult').textContent = r.to ? `${r.to.time}` : JSON.stringify(r);
  };

  // Unix
  $('#unixConvert').onclick = async () => {
    const ts = $('#unixTs').value;
    const iso = $('#unixIso').value;
    const body = ts ? { timestamp: parseFloat(ts) } : { iso };
    const r = await api('/convert/unix', { method: 'POST', body });
    $('#unixResult').innerHTML = `Epoch: ${r.timestamp}<br>ISO: ${r.iso}<br>${r.human}`;
  };

  // Duration
  $('#diffCalc').onclick = async () => {
    const r = await api('/convert/duration', { method: 'POST', body: { start: $('#diffStart').value, end: $('#diffEnd').value } });
    $('#diffResult').textContent = r.human || JSON.stringify(r);
  };

  // Date add
  $('#addCalc').onclick = async () => {
    const r = await api('/convert/date-add', { method: 'POST', body: { date: $('#addDate').value, years: parseInt($('#addY').value), months: parseInt($('#addM').value), days: parseInt($('#addD').value) } });
    $('#addResult').textContent = r.result || JSON.stringify(r);
  };

  // Age
  $('#ageCalc').onclick = async () => {
    const r = await api(`/convert/age?birthdate=${$('#ageBirth').value}`);
    $('#ageResult').textContent = r.human || JSON.stringify(r);
  };

  // Days between
  $('#betCalc').onclick = async () => {
    const r = await api(`/convert/days-between?start=${$('#betStart').value}&end=${$('#betEnd').value}`);
    $('#betResult').textContent = `${r.days} days (${(r.weeks || 0).toFixed(1)} weeks)`;
  };

  // Leap year
  $('#leapCalc').onclick = async () => {
    const r = await api(`/convert/leap-year?year=${$('#leapYear').value}`);
    $('#leapResult').textContent = r.is_leap_year ? 'Yes, leap year' : 'No, not a leap year';
  };

  // Day of week
  $('#dowCalc').onclick = async () => {
    const r = await api(`/convert/day-of-week?date=${$('#dowDate').value}`);
    $('#dowResult').textContent = r.day_of_week || JSON.stringify(r);
  };
}

export function destroy() {}
